feature(bhDebtorGroupSelect)

lomamech opened the pull request:
feature(bhDebtorGroupSelect)  …
- Implement component bhDebtorGroupSelect
- Using the bhDebtorGroupSelect in Modules: Patients, Employees, Invoice
  Registry, Cash registry,
- Update test End to End

closes #1741

lomamech committed:
feature(bhDebtorGroupSelect):
Implement components bhDebtorGroupSelect for reduce the number of
lines of code in most controllers

sfount commented on client/src/modules/cash/payments/templates/search.modal.html:
```diff
@@ -53,21 +53,11 @@
-              </ui-select-choices>
-            </ui-select>
-          </div>
+          <bh-debtor-group-select
```
:+1: This is a huge improvement for maintenance.

sfount commented on client/src/js/components/bhDebtorGroupSelect.js:
```diff
@@ -0,0 +1,51 @@
+  .component('bhDebtorGroupSelect', {
+    templateUrl : 'modules/templates/bhDebtorGroupSelect.tmpl.html',
+    controller  : DebtorGroupSelectController,
+    // transclude  : true,
```
I think it is necessary to allow this component to `transclude` to allow you to use the `bh-clear` component inside (in search modals for example).

sfount commented on client/src/modules/cash/payments/templates/search.modal.html:
```diff
@@ -53,21 +53,11 @@
+          <bh-debtor-group-select
+            debtor-group-uuid = "$ctrl.searchQueries.debtor_group_uuid"
+            on-select-callback = "$ctrl.onSelectDebtor(debtorGroup)">
+			<bh-clear on-clear="$ctrl.clear('debtor_group_uuid')"></bh-clear>
```
This inner component (`bh-clear`) is not currently displayed. I think you have to allow the component (`bh-debtor-group-select`) to `transclude`

sfount reviewed (commented):
This is a great refactor! I think there is a problem with using `transclude` on the component that should be looked at.

lomamech committed:
perf(bhDebtorGroupSelect)
Use the property transclude to all the display property bh-clear

lomamech committed:
fix(bhDebGroup): fix conflict from params of components for the test E2E

lomamech committed:
perfec(bhDebGroup)
- Update patient registry search use bhdebgroup componant
- Update test E2E for patient registry

jniles commented on client/src/js/components/bhDebtorGroupSelect.js:
```diff
@@ -0,0 +1,51 @@
+    transclude  : true,
+    bindings    : {
+      debtorGroupUuid   : '<',
+      disable           : '<?',
```
Does disable do anything?

jniles commented on client/src/js/components/bhDebtorGroupSelect.js:
```diff
@@ -0,0 +1,51 @@
+      onSelectCallback  : '&',
+      name              : '@?',
+      label             : '@?',
+      required          : '<?',
```
I don't think required does anything here.

jniles commented on client/src/modules/employees/registration/employees.js:
```diff
@@ -4,11 +4,11 @@ angular.module('bhima.controllers')
 ];
 
-function EmployeeController(Employees, Services, Grades, Functions, CreditorGroups, DebtorGroups, util, Notify, bhConstants, Receipts, Session) {
+function EmployeeController(Employees, Services, Grades, Functions, CreditorGroups, util, Notify, bhConstants, Receipts, Session,) {
```
... are you sure this works?  It looks like there should be a syntax error - the trailing comma is allowed on arrays, but not functions.

jniles commented on client/src/modules/templates/bhDebtorGroupSelect.tmpl.html:
```diff
@@ -0,0 +1,22 @@
+      ng-model="$ctrl.debtorGroupUuid"
+      on-select="$ctrl.onSelect($item, $model)" required>
+      <ui-select-match placeholder="{{ 'FORM.SELECT.DEBTOR_GROUP' | translate }}">{{$select.selected.name}}</span></ui-select-match>
+      <ui-select-choices ui-select-focus-patch repeat="group.uuid as group in ($ctrl.debtorGroups | filter:$select.search | orderBy:'name') track by group.uuid">
```
when Instead of using `| orderBy: 'name'`, it would be better to simply do the sort in the controller when you download the data - that way Angular will not have to recompute the sort on every change in the view.

jniles reviewed (approved):
@lomamech this is a good implementation.

I think there are a few pieces of code that are copied from other modules that do not belong here.  They are not blocking though.  If tests pass, LGTM :+1:
